### PR TITLE
feat: added resource scaleway_compute_instance_volume

### PR DESCRIPTION
**Pull request by DenBeke**

### CODE REVIEW

File: scaleway/resource_compute_instance_volume.go [hunk 1]
```diff
@@ -0,0 +1,155 @@
...
+				Type:        schema.TypeString,
+				Required:    true,
+				Description: "the name of the volume",
+			},
+			"size_in_gb": {
+				Type:     schema.TypeInt,
+				Required: true,
+				ValidateFunc: func(v interface{}, k string) (ws []string, errors []error) {
+					value := v.(int)
+					if value < 1 || value > 150 {
```
Comment (kindermoumoute) [inline scaleway/resource_compute_instance_volume.go]: max value is 200Go

Comment (QuentinBrosse) [reply scaleway/resource_compute_instance_volume.go]: Up @DenBeke ?

File: scaleway/resource_compute_instance_volume.go [hunk 2]
```diff
@@ -0,0 +1,203 @@
...
+				Computed:    true,
+				Description: "the name of the volume",
+				DiffSuppressFunc: func(k, old, new string, d *schema.ResourceData) bool {
+					if new == "" {
+						return true
+					}
+					return old == new
+				},
+			},
+			"size": {
```
Comment (kindermoumoute) [inline scaleway/resource_compute_instance_volume.go]: Quick suggestion: use `size_in_gb`. Ultimately the size is always in GB (I used the same naming for root_rolume on instance)

File: scaleway/resource_compute_instance_volume.go [hunk 3]
```diff
@@ -0,0 +1,203 @@
...
+			State: schema.ImportStatePassthrough,
+		},
+
+		Schema: map[string]*schema.Schema{
+			"name": {
+				Type:        schema.TypeString,
+				Optional:    true,
+				Computed:    true,
+				Description: "the name of the volume",
+				DiffSuppressFunc: func(k, old, new string, d *schema.ResourceData) bool {
```
Comment (kindermoumoute) [inline scaleway/resource_compute_instance_volume.go]: why do you need this?

File: scaleway/resource_compute_instance_volume.go [hunk 4]
```diff
@@ -0,0 +1,203 @@
...
+					if new == "" {
+						return true
+					}
+					return old == new
+				},
+			},
+			"size": {
+				Type:        schema.TypeString,
```
Comment (kindermoumoute) [inline scaleway/resource_compute_instance_volume.go]: This would be a TypeInt

File: scaleway/resource_compute_instance_volume.go [hunk 5]
```diff
@@ -0,0 +1,203 @@
...
+					}
+
+					return newInBytes == oldInBytes
+				},
+			},
+			// TODO handle snapshot, base_volume
+			"server_id": {
+				Type:        schema.TypeString,
+				Computed:    true,
+				Description: "the server associated with this volume",
```
Comment (kindermoumoute) [inline scaleway/resource_compute_instance_volume.go]: add `ValidateFunc: validationUUID(),`

File: scaleway/resource_compute_instance_volume.go [hunk 6]
```diff
@@ -0,0 +1,203 @@
...
+			},
+			"project_id": projectIDSchema(),
+			"zone":       zoneSchema(),
+		},
+	}
+}
+
+func resourceScalewayComputeInstanceVolumeCreate(d *schema.ResourceData, m interface{}) error {
+
+	meta := m.(*Meta)
```
Comment (kindermoumoute) [inline scaleway/resource_compute_instance_volume.go]: use

```go
instanceApi, zone, err := getInstanceAPIWithZone(d, m)
	if err != nil {
		return err
	}
```

File: scaleway/resource_compute_instance_volume.go [hunk 7]
```diff
@@ -0,0 +1,203 @@
...
+	}
+
+	d.SetId(newZonedId(zone, res.Volume.ID))
+
+	return resourceScalewayComputeInstanceVolumeRead(d, m)
+}
+
+func resourceScalewayComputeInstanceVolumeRead(d *schema.ResourceData, m interface{}) error {
+
+	meta := m.(*Meta)
```
Comment (kindermoumoute) [inline scaleway/resource_compute_instance_volume.go]: use
```go
	instanceApi, zone, ID, err := getInstanceAPIWithZoneAndID(m, d.Id())
	if err != nil {
		return err
	}
```

File: scaleway/resource_compute_instance_volume.go [hunk 8]
```diff
@@ -0,0 +1,203 @@
...
+	} else {
+		d.Set("server_id", nil)
+	}
+
+	return nil
+}
+
+func resourceScalewayComputeInstanceVolumeUpdate(d *schema.ResourceData, m interface{}) error {
+
+	meta := m.(*Meta)
```
Comment (kindermoumoute) [inline scaleway/resource_compute_instance_volume.go]: use
```go
	instanceApi, zone, ID, err := getInstanceAPIWithZoneAndID(m, d.Id())
	if err != nil {
		return err
	}
```

File: scaleway/resource_compute_instance_volume.go [hunk 9]
```diff
@@ -0,0 +1,203 @@
...
+		}
+
+	}
+
+	return resourceScalewayComputeInstanceVolumeRead(d, m)
+}
+
+func resourceScalewayComputeInstanceVolumeDelete(d *schema.ResourceData, m interface{}) error {
+
+	meta := m.(*Meta)
```
Comment (kindermoumoute) [inline scaleway/resource_compute_instance_volume.go]: use
```go
	instanceApi, zone, ID, err := getInstanceAPIWithZoneAndID(m, d.Id())
	if err != nil {
		return err
	}
```

File: scaleway/resource_compute_instance_volume.go [hunk 10]
```diff
@@ -0,0 +1,203 @@
...
+			d.SetId("")
+			return nil
+		}
+		return fmt.Errorf("couldn't read volume: %v", err)
+	}
+
+	d.Set("name", res.Volume.Name)
+	d.Set("size", res.Volume.Size)
+	d.Set("project_id", res.Volume.Organization)
+	d.Set("zone", string(zone))
```
Comment (kindermoumoute) [inline scaleway/resource_compute_instance_volume.go]: IMO zone shouldn't be handled here but in one of the helper function 🤔

Comment (kindermoumoute) [reply scaleway/resource_compute_instance_volume.go]: Somewhere in getZone

```go

// getZone will try to guess the zone from the following:
//  - zone field of the resource data
//  - default zone from config
func getZone(d terraformResourceData, meta *Meta) (utils.Zone, error) {

	rawZone, exist := d.GetOkExists("zone")
	if exist {
		return utils.ParseZone(rawZone.(string))
	}

	zone, exist := meta.scwClient.GetDefaultZone()
	if exist {
		d.Set("zone", string(zone)) // here
		return zone, nil
	}

	return utils.Zone(""), fmt.Errorf("could not detect region")
}
```

File: scaleway/resource_compute_instance_volume.go (hunk 3 above)
Comment (DenBeke) [reply scaleway/resource_compute_instance_volume.go]: Because the name can be auto-generated. So we need a way to check if it has changed.

Comment (kindermoumoute) [reply scaleway/resource_compute_instance_volume.go]: Let's add a function helper so we can reuse this elsewhere

File: scaleway/resource_compute_instance_volume.go (hunk 5 above)
Comment (DenBeke) [reply scaleway/resource_compute_instance_volume.go]: It's a computed field, so I cannot ad a validate function.

File: scaleway/helpers.go [hunk 11]
```diff
@@ -349,3 +349,11 @@ func regionSchema() *schema.Schema {
 func getRandomName(prefix string) string {
 	return namegenerator.GetRandomName("tf", prefix)
 }
+
+// DiffSuppressFuncForRandomName is a generic DiffSuppressFunc to avoid changing randomly generated names.
+func DiffSuppressFuncForRandomName(k, old, new string, d *schema.ResourceData) bool {
```
Comment (kindermoumoute) [inline scaleway/helpers.go]: should be private

File: scaleway/helpers.go [hunk 12]
```diff
@@ -349,3 +349,11 @@ func regionSchema() *schema.Schema {
 func getRandomName(prefix string) string {
 	return namegenerator.GetRandomName("tf", prefix)
 }
+
+// diffSuppressFuncForRandomName is a generic DiffSuppressFunc to avoid changing randomly generated names.
+func diffSuppressFuncForRandomName(k, old, new string, d *schema.ResourceData) bool {
```
Comment (QuentinBrosse) [inline scaleway/helpers.go]: nit: `RandomNameDiffSuppressFunc` ?

File: scaleway/resource_compute_instance_volume.go [hunk 13]
```diff
@@ -0,0 +1,159 @@
...
+				Description: "the server associated with this volume",
+			},
+			"project_id": projectIDSchema(),
+			"zone":       zoneSchema(),
+		},
+	}
+}
+
+func resourceScalewayComputeInstanceVolumeCreate(d *schema.ResourceData, m interface{}) error {
+
```
Comment (QuentinBrosse) [inline scaleway/resource_compute_instance_volume.go]: newline

File: scaleway/resource_compute_instance_volume.go [hunk 14]
```diff
@@ -0,0 +1,159 @@
...
+		return fmt.Errorf("couldn't create volume: %s", err)
+	}
+
+	d.SetId(newZonedId(zone, res.Volume.ID))
+
+	return resourceScalewayComputeInstanceVolumeRead(d, m)
+}
+
+func resourceScalewayComputeInstanceVolumeRead(d *schema.ResourceData, m interface{}) error {
+
```
Comment (QuentinBrosse) [inline scaleway/resource_compute_instance_volume.go]: newline

File: scaleway/resource_compute_instance_volume.go [hunk 15]
```diff
@@ -0,0 +1,159 @@
...
+	}
+
+	var (
+		volumeName = d.Get("name").(string)
+		volumeSize = uint64(d.Get("size_in_gb").(int))
+		projectID  = d.Get("project_id").(string)
+	)
+
+	// Convert human readable volume size to int in bytes
+	volumeSizeInBytes := volumeSize * gb
```
Comment (QuentinBrosse) [inline scaleway/resource_compute_instance_volume.go]: Move the `gb` definition in utils?

File: scaleway/resource_compute_instance_volume.go [hunk 16]
```diff
@@ -0,0 +1,159 @@
...
+
+	// Generate name if not set
+	if volumeName == "" {
+		volumeName = namesgenerator.GetRandomName()
```
Comment (QuentinBrosse) [inline scaleway/resource_compute_instance_volume.go]: use `getRandomName("vol")`

File: scaleway/resource_compute_instance_volume.go [hunk 17]
```diff
@@ -0,0 +1,159 @@
...
+		d.Set("server_id", res.Volume.Server.ID)
+	} else {
+		d.Set("server_id", nil)
+	}
+
+	return nil
+}
+
+func resourceScalewayComputeInstanceVolumeUpdate(d *schema.ResourceData, m interface{}) error {
+
```
Comment (QuentinBrosse) [inline scaleway/resource_compute_instance_volume.go]: newline

File: scaleway/resource_compute_instance_volume.go [hunk 18]
```diff
@@ -0,0 +1,159 @@
...
+	instanceAPI, zone, id, err := getInstanceAPIWithZoneAndID(m, d.Id())
+	if err != nil {
+		return err
+	}
+
+	if d.HasChange("name") {
+
```
Comment (QuentinBrosse) [inline scaleway/resource_compute_instance_volume.go]: newline

File: scaleway/resource_compute_instance_volume.go [hunk 19]
```diff
@@ -0,0 +1,159 @@
...
+		_, err = instanceAPI.UpdateVolume(&instance.UpdateVolumeRequest{
+			VolumeID: id,
+			Zone:     zone,
+			Name:     &newName,
+		})
+		if err != nil {
+			return fmt.Errorf("couldn't update volume: %s", err)
+		}
+
```
Comment (QuentinBrosse) [inline scaleway/resource_compute_instance_volume.go]: newline

File: scaleway/resource_compute_instance_volume.go [hunk 20]
```diff
@@ -0,0 +1,159 @@
...
+	}
+
+	if d.HasChange("name") {
+
+		newName := d.Get("name").(string)
+
+		_, err = instanceAPI.UpdateVolume(&instance.UpdateVolumeRequest{
+			VolumeID: id,
+			Zone:     zone,
+			Name:     &newName,
```
Comment (QuentinBrosse) [inline scaleway/resource_compute_instance_volume.go]: `Name: d.Get("name").(string)`

File: scaleway/resource_compute_instance_volume.go [hunk 21]
```diff
@@ -0,0 +1,159 @@
...
+			return fmt.Errorf("couldn't update volume: %s", err)
+		}
+
+	}
+
+	return resourceScalewayComputeInstanceVolumeRead(d, m)
+}
+
+func resourceScalewayComputeInstanceVolumeDelete(d *schema.ResourceData, m interface{}) error {
+
```
Comment (QuentinBrosse) [inline scaleway/resource_compute_instance_volume.go]: newline

File: scaleway/helpers.go (hunk 12 above)
Comment (DenBeke) [reply scaleway/helpers.go]: The field in TF is called `DiffSuppressFunc` so I find it logical to have it first in the name of the function. But I guess it's not that important...

**Pull request merged**